source/common/upstream/cds_api_incremental_impl.cc

```diff
@@ -0,0 +1,84 @@
+#include "common/upstream/cds_api_incremental_impl.h"
+
+#include <string>
+
+#include "envoy/api/v2/cds.pb.validate.h"
+#include "envoy/api/v2/cluster/outlier_detection.pb.validate.h"
+#include "envoy/stats/scope.h"
+
+#include "common/common/cleanup.h"
+#include "common/config/resources.h"
+#include "common/config/subscription_factory.h"
+#include "common/config/utility.h"
+#include "common/protobuf/utility.h"
+
+namespace Envoy {
+namespace Upstream {
+
+CdsApiPtr CdsApiIncrementalImpl::create(const envoy::api::v2::core::ConfigSource& cds_config,
+                                        ClusterManager& cm, Event::Dispatcher& dispatcher,
+                                        Runtime::RandomGenerator& random,
+                                        const LocalInfo::LocalInfo& local_info,
+                                        Stats::Scope& scope) {
+  return CdsApiPtr{
+      new CdsApiIncrementalImpl(cds_config, cm, dispatcher, random, local_info, scope)};
+}
+
+CdsApiIncrementalImpl::CdsApiIncrementalImpl(const envoy::api::v2::core::ConfigSource& cds_config,
+                                             ClusterManager& cm, Event::Dispatcher& dispatcher,
+                                             Runtime::RandomGenerator& random,
+                                             const LocalInfo::LocalInfo& local_info,
+                                             Stats::Scope& scope)
+    : cm_(cm), scope_(scope.createScope("cluster_manager.cds.")) {
+  Config::Utility::checkLocalInfo("cds", local_info);
+  subscription_ =
+      Config::SubscriptionFactory::subscriptionFromConfigSource<envoy::api::v2::Cluster>(
+          cds_config, local_info, dispatcher, cm, random, *scope_, "not implemented",
+          "envoy.api.v2.ClusterDiscoveryService.IncrementalClusters");
+}
+
+void CdsApiIncrementalImpl::onConfigUpdate(
+    const Protobuf::RepeatedPtrField<envoy::api::v2::Resource>& added_resources,
+    const Protobuf::RepeatedPtrField<std::string>& removed_resources,
+    const std::string& system_version_info) {
+  cm_.adsMux().pause(Config::TypeUrl::get().ClusterLoadAssignment);
+  Cleanup eds_resume([this] { cm_.adsMux().resume(Config::TypeUrl::get().ClusterLoadAssignment); });
+
+  for (const auto& resource : added_resources) {
+    try {
+      envoy::api::v2::Cluster cluster =
+          MessageUtil::anyConvert<envoy::api::v2::Cluster>(resource.resource());
+      MessageUtil::validate(cluster);
+      if (cm_.addOrUpdateCluster(cluster, resource.version())) {
+        ENVOY_LOG(debug, "cds: add/update cluster '{}'", cluster.name());
+      }
+    } catch (EnvoyException e) {
```

Comment:
I'm getting "error: catching polymorphic type 'class Envoy::EnvoyException' by value" when trying to compile. Replacing this to a reference makes the compiler happy.

Reply:
Weird; I am not getting any of these problems. I double-checked with some print statements that my runs of incremental_cds_integration_test really are executing this code. What sort of platform/compiler/C++ standard flags are you using?

Reply:
g++ (GCC) 8.2.1 on Fedora 29, debug build (using -c dbg), otherwise "normal" bazel build. It is a bit weird, esp. segfauts ...

Reply:
Also: no recent changes in the environment. This is pertinent in regards to use of insert iterators, as std::copy used to work (I think?) without them before. FWIW, I understand that it is better to use them?

Reply:
Ok, SGTM!